### ISSUE TITLE
BulkInsert include PK

### ISSUE THREAD

**Issue by tbasallo** [schotime/NPoco#558]
Looking through the source code, I think this should work, but it seem that the "Id" field is not being inserted as part of the BulkInsert.

I am trying to move data from one DB to another. Looping through the records is very slow. If I let a new ID be generated, the BulkInsert works just fine, but the ID needs to remain the same.

I have done the following:
Set AutoIncrement to false
The real primary key is ID.
I tried changing the PK to a fake column name and a real column that is not ID.

I always get the same error.

```

var data = dbOld.Fetch<T>($"SELECT * FROM {table} ORDER BY Message");
dbNew.Execute($"SET IDENTITY_INSERT {table} ON;");
dbNew.InsertBulk(data);
dbNew.Execute($"SET IDENTITY_INSERT {db} OFF;");
            
```

I then get the error:

System.Data.SqlClient.SqlException (0x80131904): Explicit value must be specified for identity column in table '[table name]' either when IDENTITY_INSERT is set to ON or when a replication user is inserting into a NOT FOR REPLICATION identity column.

**Comment by Extant-1** [comment]
This looks like it is an introduced bug in 4.0.1. In the previous version 3.9.4 this works fine.

 It seems the way bulk insert works was changed in the most recent release.

**Comment by schotime** [comment]
The sql bulk copy hasn't changed.
What does your model look like?

**Comment by tbasallo** [comment]
I tested with 5 classes and they all behaved the same. I changed AutoIncrement to true and false with no difference in behavior

```
    [TableName("{table name}"), PrimaryKey("Id", AutoIncrement = false)]
    public class Attorney
    {
        public int Id { get; set; }
        public string FirstName { get; set; } <= a bunf of prrps -
   }
```

**Comment by schotime** [comment]
The only thing that was changed was to enable it for .netstandard2.0 as that was added to it, so unless the netstandard2.0 works differently. hmmm

**Comment by tbasallo** [comment]
Just tried again. I made the class two properties - Id and Name. I tried four variations:

autoincrement on/off
identity insert on/off

This is the exact error:
```

Importing App_Attorneys @ 8/15/2019 3:23 AM
Failed on App_Attorneys - System.Data.SqlClient.SqlException (0x80131904): Explicit value must be specified for identity column in table 'App_Attorneys' either when IDENTITY_INSERT is set to ON or when a replication user is inserting into a NOT FOR REPLICATION identity column.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.TdsParser.Run(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj)
   at System.Data.SqlClient.SqlBulkCopy.RunParser(BulkCopySimpleResultSet bulkCopyHandler)
   at System.Data.SqlClient.SqlBulkCopy.CopyBatchesAsyncContinuedOnSuccess(BulkCopySimpleResultSet internalResults, String updateBulkCommandText, CancellationToken cts, TaskCompletionSource`1 source)
   at System.Data.SqlClient.SqlBulkCopy.CopyBatchesAsyncContinued(BulkCopySimpleResultSet internalResults, String updateBulkCommandText, CancellationToken cts, TaskCompletionSource`1 source)
   at System.Data.SqlClient.SqlBulkCopy.CopyBatchesAsync(BulkCopySimpleResultSet internalResults, String updateBulkCommandText, CancellationToken cts, TaskCompletionSource`1 source)
   at System.Data.SqlClient.SqlBulkCopy.WriteToServerInternalRestContinuedAsync(BulkCopySimpleResultSet internalResults, CancellationToken cts, TaskCompletionSource`1 source)
   at System.Data.SqlClient.SqlBulkCopy.WriteToServerInternalRestAsync(CancellationToken cts, TaskCompletionSource`1 source)
   at System.Data.SqlClient.SqlBulkCopy.WriteToServerInternalAsync(CancellationToken ctoken)
   at System.Data.SqlClient.SqlBulkCopy.WriteRowSourceToServerAsync(Int32 columnCount, CancellationToken ctoken)
   at System.Data.SqlClient.SqlBulkCopy.WriteToServer(DataTable table, DataRowState rowState)
   at NPoco.SqlBulkCopyHelper.BulkInsert[T](IDatabase db, IEnumerable`1 list, SqlBulkCopyOptions sqlBulkCopyOptions)
   at NPoco.Database.InsertBulk[T](IEnumerable`1 pocos)
   at etd_move.Program.Insert[T](Database dbOld, Database dbNew, String db, String primary) in C:\Projects\etd_move\etd_move\Program.cs:line 104
ClientConnectionId:c7fc0af6-00bd-48df-8310-a77ff34f37b2
Error Number:545,State:1,Class:16

```

Interestingly enough, if I remove IDENTITY from the table, then the ID will be inserted correctly whether AutoIncrement is true or false.

**Comment by tbasallo** [comment]
So, I went ahead and built locally and ran against that so that I could walk thru the code (targeting pack for 4.71!?! Downloaded, installed). And then some [docs reading](https://docs.microsoft.com/en-us/sql/relational-databases/import-export/keep-identity-values-when-bulk-importing-data-sql-server?view=sql-server-2017) and then...

_SqlBulkCopyOptions_ and then **SqlBulkCopyOptions.KeepIdentity** and then it worked.

It seems that for identity to be inserted when using bulkcopy you don't set insert identity on, you use SqlBulkCopyOptions.KeepIdentity . Currently SqlBulkCopyOptions.Default is hard coded when using the  method on Database.s. However, you can just call the helper directly and set the bulk copy option as needed:

     var db = new NPoco.Database(conn);
     NPoco.SqlBulkCopyHelper.BulkInsert({db}, data, SqlBulkCopyOptions.KeepIdentity);

Just make sure to have an open connection or call OpenSharedConnection()

I'm assuming that the reason there's no BulkOption parameter provided is becaus eonly SqlServer supports it. SO calling it like above is he only solution.

**Comment by Extant-1** [comment]
> The only thing that was changed was to enable it for .netstandard2.0 as that was added to it, so unless the netstandard2.0 works differently. hmmm

If you run a trace using SQL Server Profiler on NPoco 4.0.1 and run the db.InsertBulk method then compare the generated SQL commands to that of NPoco 3.9.4 they are different.

4.0.1 performs an actual "bulk insert" command that does not include the primary key while 3.9.4 performs an "exec sp_executesql 'INSERT INTO' command that includes the primary key.

**Comment by tbasallo** [comment]
In addition to the KeepIdentity option, AutoIncrement must be set to false since it will exclude the column when the DataTable is generated, here: https://github.com/schotime/NPoco/blob/master/src/NPoco/SqlBulkCopyHelper.cs#L48

I think it would make sense to check if KeepIdentity is set and then exclude the AutoIncrement check if it is.

@schotime would you accept that?

**Comment by schotime** [comment]
Thoughts?

```cs
var cols = pocoData.Columns.Where(x =>
            {
                if (x.Value.ResultColumn) return false;
                if (x.Value.ComputedColumn) return false;
                if (x.Value.ColumnName.Equals(pocoData.TableInfo.PrimaryKey, StringComparison.OrdinalIgnoreCase))
                {
                    if (sqlBulkCopyOptions == SqlBulkCopyOptions.KeepIdentity)
                        return true;

                    return pocoData.TableInfo.AutoIncrement == false;
                }
                return true;

            }).ToList();
```

**Comment by schotime** [comment]
If you want the old functionality just write you're own helper that does a foreach then calls instead cause thats what InsertBulk did if it wasn't supported by the platform

**Comment by tbasallo** [comment]
> Thoughts?
> 
> ```cs
> var cols = pocoData.Columns.Where(x =>
> {
>     if (x.Value.ResultColumn) return false;
>     if (x.Value.ComputedColumn) return false;
>     if (x.Value.ColumnName.Equals(pocoData.TableInfo.PrimaryKey, StringComparison.OrdinalIgnoreCase))
>     {
>         if (sqlBulkCopyOptions == SqlBulkCopyOptions.KeepIdentity)
>             return true;
> 
>         return pocoData.TableInfo.AutoIncrement == false;
>     }
>     return true;
> 
> }).ToList();
> ```

Looks good. And much easier to read than what I did!

```
var cols = pocoData.Columns.Where(x => !x.Value.ResultColumn && !x.Value.ComputedColumn
     && !(!sqlBulkCopyOptions.Equals(SqlBulkCopyOptions.KeepIdentity) && pocoData.TableInfo.AutoIncrement && x.Value.ColumnName.Equals(pocoData.TableInfo.PrimaryKey,
                                                         StringComparison.OrdinalIgnoreCase))).ToList();
```

**Comment by schotime** [comment]
Fixed in 9cbb84a3345ce1fda612d6ea6bc1bf5d775b8e7e
